src/utils/src/arg_parser.rs

```diff
@@ -0,0 +1,567 @@
+        // for valid args.
+        self.args.get(arg_name).unwrap().user_value.clone()
+    }
+
+    pub fn extra_args(self) -> Vec<String> {
+        self.extra_args
+    }
+
+    // Splits `args` in two slices: one with the actual arguments of the process and the other with
+    // the extra arguments, meaning all parameters specified after `--` (i.e. end of command options).
+    fn split_args(args: &[String]) -> (&[String], &[String]) {
+        if let Some(index) = args.iter().position(|arg| arg == ARG_SEPARATOR) {
+            return (&args[..index], &args[index + 1..]);
+        }
+
+        (&args, &[])
+    }
+
+    // Collects the command line arguments and the values provided for them.
+    pub fn parse(&mut self) -> Result<Help> {
+        let args: Vec<String> = env::args().collect();
+
+        // Skipping the first element of `args` as it is the name of the binary, not an actual argument.
+        let (args, extra_args) = ArgParser::split_args(&args[1..]);
+        self.extra_args = extra_args.to_vec();
+
+        // If user provided the `--help` argument, we want to stop the parsing.
+        if args.contains(&HELP_ARG.to_string()) {
+            return Ok(Help::Provided);
+        }
+        self.get_matches(args)?;
+        Ok(Help::NotProvided)
+    }
+
+    fn validate_matches(&self, args: &[String]) -> Result<()> {
+        for arg_info in self.args.values() {
+            // The arguments that are marked `required` must be provided by user.
+            if arg_info.required && arg_info.user_value.is_none() {
+                return Err(Error::MissingArgument(arg_info.name.to_string()));
+            }
+            // For the arguments that require a specific argument to be also present in the list
+            // of arguments provided by user, search for that argument.
+            if arg_info.user_value.is_some() {
+                if let Some(arg_name) = arg_info.requires {
+                    if !args.contains(&(format!("--{}", arg_name))) {
+                        return Err(Error::MissingArgument(arg_name.to_string()));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_arg(&self, arg: &str) -> Result<()> {
+        if !arg.starts_with(ARG_SEPARATOR) {
+            return Err(Error::UnexpectedArgument(arg.to_string()));
+        }
+        let arg_name = &arg[ARG_SEPARATOR.len()..];
+
+        // Check if the argument is an expected one.
+        if !self.args.contains_key(arg_name) {
+            return Err(Error::UnexpectedArgument(arg_name.to_string()));
+        }
+
+        // Check if the argument was not provided more than once.
+        if self.args.get(arg_name).unwrap().user_value.is_some() {
+            return Err(Error::DuplicateArgument(arg_name.to_string()));
+        }
+        Ok(())
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(&mut self, args: &[String]) -> Result<()> {
```

Comment:
A `get_` function implies a returned value, but this function doesn't have one. This function looks like the workhorse of `ArgParser`, while `ArgParser::parse()` deals with minor setup and calls into here.

I don't really understand the use of `matches` either here or with `validate_matches`. Can you please clarify this either by choosing more specific names, or by commenting in a short description of how this all works?